npe2/_inspection/_visitors.py

```diff
@@ -0,0 +1,455 @@
+
+
+class NPE2PluginModuleVisitor(_DecoratorVisitor):
+    """AST visitor to find all contributions in an npe2 module.
+
+    This visitor will find all the contributions (things decorated with
+    `@npe2.implements`) in a module and store them in `contribution_points`.
+
+    It works as follows:
+
+    See how decorators are found in the docstring for `_DecoratorVisitor`.
+    when a decorator is found that matches `npe2.implements` (e.g. `@implements.reader`)
+
+    4. `_store_contrib` first calls `_store_command` which does the job of storing the
+       fully-resolved `python_name` for the function being decorated, and creates a
+       CommandContribution. `_store_contrib` will then create the appropriate
+       contribution type (e.g. `npe2.implements.reader` will instantiate a
+       `ReaderContribution`), set its `command` entry to the `id` of the just-created
+       `CommandContribution`, then store it in `contribution_points`.
+    5. When the visitor is finished, we can create an instance of `ContributionPoints`
+       using `ContributionPoints(**visitor.contribution_points)`, then add it to a
+       manifest.
+    """
+
+    def __init__(
+        self, plugin_name: str, module_name: str, match: str = "npe2.implements"
+    ) -> None:
+        super().__init__(module_name, match)
+        self.plugin_name = plugin_name
+        self.contribution_points: Dict[str, List[dict]] = {}
+
+    def _process_decorated(
+        self,
+        decorator_name: str,
+        node: Union[ast.ClassDef, ast.FunctionDef],
+        decorator_kwargs: dict,
+    ):
+        self._store_contrib(decorator_name, node.name, decorator_kwargs)
+
+    def _store_contrib(self, contrib_type: str, name: str, kwargs: Dict[str, Any]):
+        from ..implements import CHECK_ARGS_PARAM  # circ import
+
+        kwargs.pop(CHECK_ARGS_PARAM, None)
+        ContribClass, contrib_name = CONTRIB_MAP[contrib_type]
+        contrib = ContribClass(**self._store_command(name, kwargs))
+        existing: List[dict] = self.contribution_points.setdefault(contrib_name, [])
+        existing.append(contrib.dict(exclude_unset=True))
+
+    def _store_command(self, name: str, kwargs: Dict[str, Any]) -> Dict[str, Any]:
+
+        cmd_kwargs = {k: kwargs.pop(k) for k in list(kwargs) if k in _COMMAND_PARAMS}
+        cmd_kwargs["python_name"] = self._qualified_pyname(name)
+        cmd = contributions.CommandContribution(**cmd_kwargs)
+        if cmd.id.startswith(self.plugin_name):
+            n = len(self.plugin_name)
+            cmd.id = cmd.id[n:]
+        cmd.id = f"{self.plugin_name}.{cmd.id.lstrip('.')}"
+        cmd_contribs: List[dict] = self.contribution_points.setdefault("commands", [])
+        cmd_contribs.append(cmd.dict(exclude_unset=True))
+        kwargs["command"] = cmd.id
+        return kwargs
+
+    def _qualified_pyname(self, obj_name: str) -> str:
+        return f"{self.module_name}:{obj_name}"
+
+
+class NPE1PluginModuleVisitor(_DecoratorVisitor):
```

Comment:
this is new, but it closely follows the logic in the hookimpl visitor in _from_npe1 ... just with ast instead of dynamic code